app/models/concerns/reviewable_action_builder.rb

```diff
@@ -241,6 +241,101 @@ def perform_convert_to_pm(performed_by, _args)
     create_result(:success, :rejected, [created_by_id], false)
   end
 
+  # Calculate the final reviewable status based on the action logs. This method
+  # assumes that all_bundles_actioned? has already returned true.
+  #
+  # @return [Symbol] The calculated final status (:ignored, :rejected, :approved, or :pending)
+  def calculate_final_status_from_logs
+    statuses = reviewable_action_logs.pluck(:status).uniq
+
+    return :pending if statuses.empty?
+
+    ignored_value = Reviewable.statuses["ignored"]
+    rejected_value = Reviewable.statuses["rejected"]
+    approved_value = Reviewable.statuses["approved"]
+
+    return :ignored if statuses.all? { |s| s == "ignored" || s == ignored_value }
+    return :rejected if statuses.all? { |s| s == "rejected" || s == rejected_value }
+    return :approved if statuses.any? { |s| s == "approved" || s == approved_value }
+
+    :pending
+  end
+
+  # Check if all action bundles have been addressed with at least one action.
+  #
+  # @param guardian [Guardian] The guardian to check permissions for.
+  # @param args [Hash] Additional arguments for building actions.
+  #
+  # @return [Boolean] True if all bundles have at least one logged action.
+  def all_bundles_actioned?(guardian, args = {})
+    actions = actions_for(guardian, args)
+    # Extract bundle types from bundle IDs (e.g., "8311-post-actions" -> "post-actions")
+    current_bundle_types = actions.bundles.map { |b| b.id.split("-", 2).last }
+    logged_bundle_types = reviewable_action_logs.pluck(:bundle).uniq.compact
+
+    # Check if at least one action from each bundle type has been logged
+    current_bundle_types.all? { |type| logged_bundle_types.include?(type) }
+  end
+
+  # Override the Reviewable#perform method to add action logging, and to handle deferred finalization
```

Comment:
Instead of override, I think we can modify original method and add small
```ruby
use_deferred_transitions =
      guardian.can_see_reviewable_ui_refresh? && !SiteSetting.reviewable_old_moderator_actions

if use_deferred_transions
...
else
...
end
```